### PR TITLE
feat: Buildkite pipeline for building dev docker images [ES-9318]

### PR DESCRIPTION
**Pull request by favilo**

### CODE REVIEW

File: release-docker-manifest.sh
Comment (favilo) [inline release-docker-manifest.sh:42]: NOTE: need to also push to `${RALLY_VERSION}`

File: release-docker.sh
Comment (favilo) [inline release-docker.sh:64]: I want this bit to be conditional. What if we're building an older version?

File: build-dev-docker.sh
Comment (favilo) [inline build-dev-docker.sh:55]: Before we build, we need to check out the `BUILD_FROM_BRANCH` branch and build _that_ source from our current `Dockerfile`

File: .buildkite/dev-docker/run.sh
Comment (favilo) [inline .buildkite/dev-docker/run.sh:9]: If empty, use `BUILDKITE_BRANCH`

File: build-dev-docker.sh
Comment (favilo) [reply build-dev-docker.sh:62]: Checked out into `$rally_dir`

File: build-dev-docker.sh
Comment (favilo) [inline build-dev-docker.sh:45]: @gbanasiak Is this fine? Or should we move it somewhere else?

Comment (gbanasiak) [reply build-dev-docker.sh:45]: I think we should aim for `docker.elastic.co/rally` but can start from that.

File: build-dev-docker.sh
Comment (gbanasiak) [reply build-dev-docker.sh:62]: So this creates the new temporary directory and clones Rally, but where's the checkout?

File: .buildkite/dev-docker/manifest.sh
Comment (gbanasiak) [inline .buildkite/dev-docker/manifest.sh:19]: We're missing Elastic Docker registry variant here. Would be good to avoid repetition with `run.sh` too, maybe a shared file, like `retry.sh` ?

File: .buildkite/dev-docker/run.sh
Comment (gbanasiak) [inline .buildkite/dev-docker/run.sh:35]:
```suggestion
retry 5 docker login -u ${DOCKER_USERNAME} -p ${DOCKER_PASSWORD} ${DOCKER_REGISTRY}
```

File: build-dev-docker.sh
Comment (favilo) [reply build-dev-docker.sh:45]: Alright, I'll leave it for now, until we get the TLD permissions to push there.

File: .buildkite/dev-docker/manifest.sh
Comment (favilo) [reply .buildkite/dev-docker/manifest.sh]: Good idea, I've merged both `run.sh` and `manifest.sh` into a single script

File: .buildkite/dev-docker/run.sh
Comment (gbanasiak) [inline .buildkite/dev-docker/run.sh:47]: I like the idea behind `git --no-pager show`, but this is potentially quite verbose. How about changing this to something a bit more brief, e.g. `git log --oneline -n1`? I would also add comments corresponding to the context where we checkout:

```
echo "Dockerfile commit: $(git log --oneline -n1)"
```

vs.

```
echo "Rally commit: $(git log --oneline -n1)"
```

**Review by gbanasiak — APPROVED** [review]
Thanks for iterating, LGTM.

File: .buildkite/dev-docker/run.sh
Comment (favilo) [reply .buildkite/dev-docker/run.sh:47]: This is a great idea. I'll do that and merge. Thanks!

File: .buildkite/release-docker/run.sh
Comment (gbanasiak) [inline .buildkite/release-docker/run.sh:36]:
```suggestion
    echo "Docker commit: $(git --no-pager log --oneline -n1)"
```

File: .buildkite/dev-docker/run.sh
Comment (gbanasiak) [inline .buildkite/dev-docker/run.sh:32]: How about adding retries here like with `docker login` ?

**Comment by favilo** [comment]
Waiting for our docker user to get write access to a top level directory before I can merge this.

**Comment by favilo** [comment]
This is currently blocked on [SRERE-408]

**Comment by favilo** [comment]
> This is currently blocked on [SRERE-408]
> 
> [SRERE-408]: https://elasticco.atlassian.net/browse/SRERE-408?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

We have permissions, and the buildkite job succeeds pushing to the internal repo

**Pull request merged**
